### PR TITLE
Issue #13758: Add pattern array converter for auto bean

### PR DESCRIPTION
**Pull request by Lmh-java**
For Issue #13758.

1. Create auto configuration for `Pattern[]` with a fixed separator (in this PR)
2. Update `AbstractClassCouplingCheck` to use this new property type (in this PR)
3. Update `ImportOrderCheck` to replace use of `Pattern[]` with `String[]` (In this PR)
------

Agreement on update is archived at https://github.com/checkstyle/checkstyle/pull/14764#issuecomment-2041485231

### CODE REVIEW

**Comment by Lmh-java** [comment]
https://github.com/checkstyle/checkstyle/pull/14740#issue-2216307739
> In this PR, I have:
> 1. Create auto configuration for Pattern[] .....

@rnveach @romani Could you please share your thoughts on this plan? I truly appreciate your opinions.

Also, for https://github.com/checkstyle/checkstyle/issues/13758#issuecomment-2025700503, which one would be better in our case?

File: src/main/java/com/puppycrawl/tools/checkstyle/AbstractAutomaticBean.java [hunk 1]
```diff
@@ -313,6 +317,22 @@ public Object convert(Class type, Object value) {
...
     }
 
+    /** A converter that converts a string array to a pattern. */
+    private static final class PatternArrayConverter implements Converter {
+
+        @SuppressWarnings("unchecked")
+        @Override
+        public Object convert(Class type, Object value) {
+            final String plainString = value.toString();
+            final String[] patternStrings = plainString.split(REGEX_SEPARATOR);
```
Comment (romani) [inline src/main/java/com/puppycrawl/tools/checkstyle/AbstractAutomaticBean.java:327]: bad side effect of this is we should clearly mention that we do not allow `,` as part of regexp to match what user needs. So user can not use `,` in it Patterns, it is delimiter.
`\,` will also not work, as we simply splitting for just `,`.

**Review by romani — CHANGES_REQUESTED** [review]
item to disucuss

File: src/main/java/com/puppycrawl/tools/checkstyle/AbstractAutomaticBean.java (hunk 1 above)
Comment (Lmh-java) [reply src/main/java/com/puppycrawl/tools/checkstyle/AbstractAutomaticBean.java:327]: @romani, do you want me to close this PR for now? Since it might be confusing for others. We can re-open it later to further modify `Pattern[]` if it is the right time. :)

Comment (romani) [reply src/main/java/com/puppycrawl/tools/checkstyle/AbstractAutomaticBean.java:327]: Let's keep it open until we come to agreement on how to fix issue.

**Comment by romani** [comment]
@Lmh-java , please put in this PR all from https://github.com/checkstyle/checkstyle/pull/14764#issuecomment-2041485231

File: config/linkcheck-suppressions.txt [hunk 2]
```diff
@@ -35,6 +35,7 @@
 <a href="AbstractNode.html#getDeclaredNamespaces(net.sf.saxon.om.NamespaceBinding%5B%5D)">AbstractNode.html#getDeclaredNamespaces(net.sf.saxon.om.NamespaceBinding%5B%5D)</a>: doesn't exist.
 <a href="apidocs/com/puppycrawl/tools/checkstyle/AbstractAutomaticBean.OutputStreamOptions.html#%3Cinit%3E()">#%3Cinit%3E()</a>: doesn't exist.
 <a href="apidocs/com/puppycrawl/tools/checkstyle/AbstractAutomaticBean.PatternConverter.html#%3Cinit%3E()">#%3Cinit%3E()</a>: doesn't exist.
+<a href="apidocs/com/puppycrawl/tools/checkstyle/AbstractAutomaticBean.PatternArrayConverter.html#%3Cinit%3E()">#%3Cinit%3E()</a>: doesn't exist.
```
Comment (Lmh-java) [inline config/linkcheck-suppressions.txt:38]: Suppress link check as other converters.

File: config/pitest-suppressions/pitest-imports-suppressions.xml [hunk 3]
```diff
@@ -1,5 +1,23 @@
...
+
+  <mutation unstable="false">
+    <sourceFile>ImportOrderCheck.java</sourceFile>
+    <mutatedClass>com.puppycrawl.tools.checkstyle.checks.imports.ImportOrderCheck</mutatedClass>
+    <mutatedMethod>setStaticGroups</mutatedMethod>
+    <mutator>org.pitest.mutationtest.engine.gregor.mutators.experimental.ArgumentPropagationMutator</mutator>
+    <description>replaced call to java/util/Arrays::copyOf with argument</description>
+    <lineContent>staticGroups = Arrays.copyOf(packageGroups, packageGroups.length);</lineContent>
+  </mutation>
+
```
Comment (Lmh-java) [inline config/pitest-suppressions/pitest-imports-suppressions.xml:20]: Need to use `Arrays.copyOf()` to copy the parameter of the function and store it, otherwise PWD will complain.

File: config/pmd.xml [hunk 4]
```diff
@@ -316,7 +316,8 @@
       <!-- Main has an annotated field for each command line option. This is by design. -->
       <property name="violationSuppressXPath"
                 value="//ClassOrInterfaceDeclaration[@SimpleName='Checker']
-    | //ClassOrInterfaceDeclaration[@SimpleName='Main']"/>
+    | //ClassOrInterfaceDeclaration[@SimpleName='Main']
+    | //ClassOrInterfaceDeclaration[@SimpleName='ImportOrderCheck']"/>
```
Comment (Lmh-java) [inline config/pmd.xml:320]: Suppress this because PWD has too many class fields.

File: src/main/java/com/puppycrawl/tools/checkstyle/checks/imports/ImportOrderCheck.java [hunk 5]
```diff
@@ -262,6 +264,16 @@ public class ImportOrderCheck
...
+    /**
+     * Complied array of patterns for property {@code groups}.
+     */
+    private Pattern[] groupsReg = EMPTY_PATTERN_ARRAY;
+
+    /**
+     * Complied array of patterns for property {@code staticGroups}.
+     */
+    private Pattern[] staticGroupsReg = EMPTY_PATTERN_ARRAY;
+
```
Comment (Lmh-java) [inline src/main/java/com/puppycrawl/tools/checkstyle/checks/imports/ImportOrderCheck.java:276]: Add these two more fields to store the compiled regex expressions. I have to expose `String[]` to users, but we also need to store the complied regex (we don't want to compile every time we want to use the regex). I am curious whether there are better solutions to this, since this change brings us more suppressions of all kinds.

File: src/test/resources/com/puppycrawl/tools/checkstyle/checks/metrics/classfanoutcomplexity/InputClassFanOutComplexityLambdaNew.java [hunk 6]
```diff
@@ -0,0 +1,40 @@
...
+import java.util.List;
+import java.util.Map;
+
+class InputClassFanOutComplexityLambdaNew {
+    public void testMethod(Map<String, List<String>> entry) {
+        List<Constraint> constraints = new ArrayList<>();
+        String string = "";
+        entry.values().stream()
+                .flatMap(List::stream)
+                .map(Constraint::new)
```
Comment (Lmh-java) [inline src/test/resources/com/puppycrawl/tools/checkstyle/checks/metrics/classfanoutcomplexity/InputClassFanOutComplexityLambdaNew.java:32]: `Constraint::new` will not be recognized as a reference and will not be recorded. I am adding this test file to pass CI right now. I think it would be better if we could fix this false negative in a separated PR to make everything easier.

File: src/test/java/com/puppycrawl/tools/checkstyle/checks/metrics/ClassFanOutComplexityCheckTest.java [hunk 7]
```diff
@@ -371,4 +371,18 @@ public void testClearStatePackageName() throws Exception {
...
+     * normal declaration of a new instance. We need to make sure this does not
+     *
+     * @throws Exception when code tested throws exception
+     */
+    @Test
+    public void testLambdaNew() throws Exception {
+        final String[] expected = {};
+
+        verifyWithInlineConfigParser(
+                getPath("InputClassFanOutComplexityLambdaNew.java"), expected);
```
Comment (Lmh-java) [inline src/test/java/com/puppycrawl/tools/checkstyle/checks/metrics/ClassFanOutComplexityCheckTest.java:385]: This test case is not correct, but I have to add it to pass the current CI. `InputClassFanOutComplexityLambdaNew.java` should NOT have 0 complexity

File: config/pitest-suppressions/pitest-imports-suppressions.xml [hunk 8]
```diff
@@ -1,5 +1,23 @@
 <?xml version="1.0" encoding="UTF-8"?>
 <suppressedMutations>
+  <mutation unstable="false">
+    <sourceFile>ImportOrderCheck.java</sourceFile>
+    <mutatedClass>com.puppycrawl.tools.checkstyle.checks.imports.ImportOrderCheck</mutatedClass>
+    <mutatedMethod>setGroups</mutatedMethod>
+    <mutator>org.pitest.mutationtest.engine.gregor.mutators.experimental.ArgumentPropagationMutator</mutator>
+    <description>replaced call to java/util/Arrays::copyOf with argument</description>
+    <lineContent>groups = Arrays.copyOf(packageGroups, packageGroups.length);</lineContent>
```
Comment (romani) [inline config/pitest-suppressions/pitest-imports-suppressions.xml:9]: please use UnmodifiableCollectionUtil to avoid this.
pitest suppresison can not be extended.

**Review by romani — CHANGES_REQUESTED** [review]
items:

File: config/pitest-suppressions/pitest-imports-suppressions.xml (hunk 8 above)
Comment (Lmh-java) [reply config/pitest-suppressions/pitest-imports-suppressions.xml]: done

File: src/test/java/com/puppycrawl/tools/checkstyle/checks/imports/ImportOrderCheckTest.java [hunk 9]
```diff
@@ -795,6 +795,27 @@ public void testTrimOption() throws Exception {
...
+        checkObj.setGroups(expectedGroups);
+        checkObj.setStaticGroups(expectedStaticGroups);
+
+        final String[] groups = TestUtil.getInternalState(checkObj, "groups");
+        final String[] staticGroups = TestUtil.getInternalState(checkObj, "staticGroups");
+        groups[0] = "modified";
+        staticGroups[0] = "modified";
+        assertWithMessage("Property 'groups' is not unmodifiable")
+                .that(expectedGroups[0])
+                .isNotEqualTo("modified");
```
Comment (romani) [inline src/test/java/com/puppycrawl/tools/checkstyle/checks/imports/ImportOrderCheckTest.java:813]: if Check going to have not Input based test, it should have comment to explain why it can not be covered by Input based test, it should be be very good reason.
if possible to be Input based, must be Input based.

File: src/test/java/com/puppycrawl/tools/checkstyle/AbstractAutomaticBeanTest.java [hunk 10]
```diff
@@ -278,6 +280,46 @@ public void testBeanConvertersUri3() {
         }
     }
 
+    @Test
+    public void testBeanConverterPatternArray1() throws Exception {
```
Comment (romani) [inline src/test/java/com/puppycrawl/tools/checkstyle/AbstractAutomaticBeanTest.java:284]: `testBeanConverterPatternArray`

File: src/test/java/com/puppycrawl/tools/checkstyle/AbstractAutomaticBeanTest.java [hunk 11]
```diff
@@ -278,6 +280,46 @@ public void testBeanConvertersUri3() {
...
+        final ConverterBean bean = new ConverterBean();
+        final DefaultConfiguration config = new DefaultConfiguration("bean");
+        final String patternString = "^a*$,^b*$, ^c*$";
+        final List<String> expectedPatternString = Arrays.asList("^a*$", "^b*$", " ^c*$");
```
Comment (romani) [inline src/test/java/com/puppycrawl/tools/checkstyle/AbstractAutomaticBeanTest.java:288]: `expectedPatterns` and make conversion to Petterns here as you did in assert

File: src/test/java/com/puppycrawl/tools/checkstyle/AbstractAutomaticBeanTest.java [hunk 12]
```diff
@@ -278,6 +280,46 @@ public void testBeanConvertersUri3() {
...
+        config.addProperty("patterns", patternString);
+        bean.configure(config);
+
+        assertWithMessage("invalid result")
```
Comment (romani) [inline src/test/java/com/puppycrawl/tools/checkstyle/AbstractAutomaticBeanTest.java:292]: invalid size of result

File: src/test/java/com/puppycrawl/tools/checkstyle/AbstractAutomaticBeanTest.java [hunk 13]
```diff
@@ -278,6 +280,46 @@ public void testBeanConvertersUri3() {
...
+                .that(bean.patterns)
+                .hasLength(3);
+        assertWithMessage("invalid result")
+                .that(expectedPatternString)
+                .containsExactlyElementsIn(
+                        Arrays.stream(bean.patterns)
+                                .map(Pattern::pattern)
+                                .collect(Collectors.toUnmodifiableList()));
```
Comment (romani) [inline src/test/java/com/puppycrawl/tools/checkstyle/AbstractAutomaticBeanTest.java:300]: move actual list of patterns to local variable `actualPatterns`

File: src/test/java/com/puppycrawl/tools/checkstyle/AbstractAutomaticBeanTest.java [hunk 14]
```diff
@@ -278,6 +280,46 @@ public void testBeanConvertersUri3() {
...
+    }
+
+    @Test
+    public void testBeanConverterPatternArray2() throws Exception {
```
Comment (romani) [inline src/test/java/com/puppycrawl/tools/checkstyle/AbstractAutomaticBeanTest.java:304]: `testBeanConverterPatternArraySingleElement`

and please do make test for empty value `""` for list of patterns.

**Review by romani — CHANGES_REQUESTED** [review]
please apply review items to all cases:

File: config/pmd.xml (hunk 4 above)
Comment (romani) [reply config/pmd.xml:320]: I am ok with this.

File: src/main/java/com/puppycrawl/tools/checkstyle/checks/imports/ImportOrderCheck.java (hunk 5 above)
Comment (romani) [reply src/main/java/com/puppycrawl/tools/checkstyle/checks/imports/ImportOrderCheck.java:276]: to avoid dealing with long discussion on what is actually a type that user should care, I am ok with this update.
I hope in some future we will agree on type system and simplify this Check

File: src/test/java/com/puppycrawl/tools/checkstyle/checks/imports/ImportOrderCheckTest.java (hunk 9 above)
Comment (Lmh-java) [reply src/test/java/com/puppycrawl/tools/checkstyle/checks/imports/ImportOrderCheckTest.java:813]: I've tried to make this input-based, but I could not. This test is to kill Pitest. Pitest will automatically remove the function call to `UnmodifiableCollectionUtil.copyOfArray()` and there is no difference after such mutation. The only way this mutation can be detected is through changing the reference (as above), which cannot be done through input files.

Comment (romani) [reply src/test/java/com/puppycrawl/tools/checkstyle/checks/imports/ImportOrderCheckTest.java:813]: What this test try to cover? Pitest mutation only?

Comment (Lmh-java) [reply src/test/java/com/puppycrawl/tools/checkstyle/checks/imports/ImportOrderCheckTest.java:813]: Yes, only for covering Pitest mutation.

@romani I have some issues with Pitest running locally, I need some more time to fix it. But just for your quick reference, this test case is to cover the following two usages of `UnmodifiableCollectionUtil.copyOfArray()` in `ImportOrderCheck.java`.

```java
public void setGroups(String... packageGroups) {
        groups = UnmodifiableCollectionUtil.copyOfArray(packageGroups, packageGroups.length);
        groupsReg = compilePatterns(packageGroups);
}
```

```java
public void setStaticGroups(String... packageGroups) {
        staticGroups = UnmodifiableCollectionUtil.copyOfArray(packageGroups, packageGroups.length);
        staticGroupsReg = compilePatterns(packageGroups);
}
```

As mentioned in https://github.com/checkstyle/checkstyle/pull/14740#discussion_r1564228923, Pitest will remove the call to `UnmodifiableCollectionUtil.copyOfArray()` and complains no unit test detects this removal.

Comment (romani) [reply src/test/java/com/puppycrawl/tools/checkstyle/checks/imports/ImportOrderCheckTest.java:813]: I can not allow this, we did a lot of efforts to kill all pitest mutations by Inputs.
Please share full mutation survival block or point to CI execution when you detected it, to let me see all details.

Comment (romani) [reply src/test/java/com/puppycrawl/tools/checkstyle/checks/imports/ImportOrderCheckTest.java:813]: Please add https://github.com/checkstyle/checkstyle/blob/69337c7913430eea2e6618046e50102f770ab68d/pom.xml#L3072
To https://github.com/checkstyle/checkstyle/blob/69337c7913430eea2e6618046e50102f770ab68d/pom.xml#L3635

File: src/test/java/com/puppycrawl/tools/checkstyle/AbstractAutomaticBeanTest.java (hunk 14 above)
Comment (Lmh-java) [reply src/test/java/com/puppycrawl/tools/checkstyle/AbstractAutomaticBeanTest.java]: done

File: src/test/java/com/puppycrawl/tools/checkstyle/AbstractAutomaticBeanTest.java (hunk 13 above)
Comment (Lmh-java) [reply src/test/java/com/puppycrawl/tools/checkstyle/AbstractAutomaticBeanTest.java]: done

File: src/test/java/com/puppycrawl/tools/checkstyle/AbstractAutomaticBeanTest.java (hunk 12 above)
Comment (Lmh-java) [reply src/test/java/com/puppycrawl/tools/checkstyle/AbstractAutomaticBeanTest.java]: done

File: src/test/java/com/puppycrawl/tools/checkstyle/AbstractAutomaticBeanTest.java (hunk 11 above)
Comment (Lmh-java) [reply src/test/java/com/puppycrawl/tools/checkstyle/AbstractAutomaticBeanTest.java]: done

File: src/test/java/com/puppycrawl/tools/checkstyle/AbstractAutomaticBeanTest.java (hunk 10 above)
Comment (Lmh-java) [reply src/test/java/com/puppycrawl/tools/checkstyle/AbstractAutomaticBeanTest.java]: done

File: src/test/java/com/puppycrawl/tools/checkstyle/checks/imports/ImportOrderCheckTest.java (hunk 9 above)
Comment (Lmh-java) [reply src/test/java/com/puppycrawl/tools/checkstyle/checks/imports/ImportOrderCheckTest.java]: Sure, I've changed `pom.xml` and pushed again. I will check the Pitest result by CI.

File: src/test/resources/com/puppycrawl/tools/checkstyle/checks/metrics/classfanoutcomplexity/InputClassFanOutComplexityLambdaNew.java (hunk 6 above)
Comment (romani) [reply src/test/resources/com/puppycrawl/tools/checkstyle/checks/metrics/classfanoutcomplexity/InputClassFanOutComplexityLambdaNew.java:32]: Please use here File class, as you did in issue description

File: src/main/java/com/puppycrawl/tools/checkstyle/AbstractAutomaticBean.java [hunk 15]
```diff
@@ -313,6 +317,25 @@ public Object convert(Class type, Object value) {
...
+        @SuppressWarnings("unchecked")
+        @Override
+        public Object convert(Class type, Object value) {
+            final String plainString = value.toString();
+            Pattern[] patterns = new Pattern[0];
+            if (!CommonUtil.isBlank(plainString)) {
+                final String[] patternStrings = plainString.split(REGEX_SEPARATOR);
+                patterns = new Pattern[patternStrings.length];
+                for (int index = 0; index < patternStrings.length; index++) {
+                    patterns[index] = CommonUtil.createPattern(patternStrings[index]);
```
Comment (romani) [inline src/main/java/com/puppycrawl/tools/checkstyle/AbstractAutomaticBean.java:332]: Do we need index variable here ? Can we loop in form of for-each.

**Review by romani — CHANGES_REQUESTED** [review]
Items

File: src/test/java/com/puppycrawl/tools/checkstyle/checks/metrics/ClassFanOutComplexityCheckTest.java (hunk 7 above)
Comment (romani) [reply src/test/java/com/puppycrawl/tools/checkstyle/checks/metrics/ClassFanOutComplexityCheckTest.java:385]: Please put comment, in code, `0 violations until #14787`

File: src/main/java/com/puppycrawl/tools/checkstyle/AbstractAutomaticBean.java (hunk 15 above)
Comment (Lmh-java) [reply src/main/java/com/puppycrawl/tools/checkstyle/AbstractAutomaticBean.java]: done, replaced with for-each

File: src/test/resources/com/puppycrawl/tools/checkstyle/checks/metrics/classfanoutcomplexity/InputClassFanOutComplexityLambdaNew.java (hunk 6 above)
Comment (Lmh-java) [reply src/test/resources/com/puppycrawl/tools/checkstyle/checks/metrics/classfanoutcomplexity/InputClassFanOutComplexityLambdaNew.java]: done

File: src/test/java/com/puppycrawl/tools/checkstyle/checks/metrics/ClassFanOutComplexityCheckTest.java (hunk 7 above)
Comment (Lmh-java) [reply src/test/java/com/puppycrawl/tools/checkstyle/checks/metrics/ClassFanOutComplexityCheckTest.java]: done

File: src/test/java/com/puppycrawl/tools/checkstyle/AbstractAutomaticBeanTest.java [hunk 16]
```diff
@@ -278,6 +280,39 @@ public void testBeanConvertersUri3() {
...
+                .that(actualPatternStrings)
+                .containsExactlyElementsIn(expectedPatternStrings);
+    }
+
+    @Test
+    public void testBeanConverterPatternArraySingleElement() throws Exception {
+        final ConverterBean bean = new ConverterBean();
+        final DefaultConfiguration config = new DefaultConfiguration("bean");
+        config.addProperty("patterns", "");
+        bean.configure(config);
```
Comment (romani) [inline src/test/java/com/puppycrawl/tools/checkstyle/AbstractAutomaticBeanTest.java:309]: testBeanConverterPatternArraySingleElement

Empty is not a single, please make 2 tests, for single pattern and separate for empty.

**Review by romani — CHANGES_REQUESTED** [review]
Items

File: src/test/java/com/puppycrawl/tools/checkstyle/AbstractAutomaticBeanTest.java (hunk 16 above)
Comment (Lmh-java) [reply src/test/java/com/puppycrawl/tools/checkstyle/AbstractAutomaticBeanTest.java:330]: done

**Review by romani — APPROVED** [review]
Ok to merge

File: src/main/java/com/puppycrawl/tools/checkstyle/AbstractAutomaticBean.java [hunk 17]
```diff
@@ -84,6 +85,9 @@ public enum OutputStreamOptions {
     /** Comma separator for StringTokenizer. */
     private static final String COMMA_SEPARATOR = ",";
 
+    /** Separator that splits multiple regex expression in a string. */
+    private static final String REGEX_SEPARATOR = COMMA_SEPARATOR;
```
Comment (rnveach) [inline src/main/java/com/puppycrawl/tools/checkstyle/AbstractAutomaticBean.java:89]: I don't see the need for a field to exactly duplicate another field. Just remove it and use `COMMA_SEPARATOR`.

File: src/main/java/com/puppycrawl/tools/checkstyle/AbstractAutomaticBean.java [hunk 18]
```diff
@@ -313,6 +318,24 @@ public Object convert(Class type, Object value) {
 
     }
 
+    /** A converter that converts a string array to a pattern. */
```
Comment (rnveach) [inline src/main/java/com/puppycrawl/tools/checkstyle/AbstractAutomaticBean.java:321]: > that converts a string array 

Isn't it just converting a string and not a string array?

File: src/xdocs/property_types.xml [hunk 19]
```diff
@@ -673,7 +673,7 @@
     <section name="Pattern[]">
       <p>
         This type represents a set of patterns. The string representation
-        is parsed as a set of forward slash ('/') separated patterns.
+        is parsed as a set of comma (',') separated patterns.
```
Comment (rnveach) [inline src/xdocs/property_types.xml:676]: Add this text that can be seen on other properties.

> Alternatively, a property of this type can be supplied multiple times which is equivalent to a set of comma separated patterns

File: src/test/java/com/puppycrawl/tools/checkstyle/checks/metrics/ClassFanOutComplexityCheckTest.java [hunk 20]
```diff
@@ -371,4 +371,18 @@ public void testClearStatePackageName() throws Exception {
                 .isTrue();
     }
 
+    /**
+     * Test [ClassName]::new expression. Such expression will be processed as a
```
Comment (rnveach) [inline src/test/java/com/puppycrawl/tools/checkstyle/checks/metrics/ClassFanOutComplexityCheckTest.java:375]: What does this have to do with the pattern array converter? All the properties in the file are default which means it doesn't go against the converter, right?

File: src/main/resources/com/puppycrawl/tools/checkstyle/meta/checks/imports/ImportOrderCheck.xml [hunk 21]
```diff
@@ -40,7 +40,7 @@
  &lt;a href="https://en.wikipedia.org/wiki/ASCII#Order"&gt;ASCII sort order&lt;/a&gt;.
  It affects both type imports and static imports.</description>
             </property>
-            <property default-value="" name="groups" type="java.util.regex.Pattern[]">
+            <property default-value="" name="groups" type="java.lang.String[]">
```
Comment (rnveach) [inline src/main/resources/com/puppycrawl/tools/checkstyle/meta/checks/imports/ImportOrderCheck.xml:43]: Has it been discussed what impact this has on 3rd party tools like eclipse-cs?

**Review by rnveach — CHANGES_REQUESTED** [review]
> Start Issue https://github.com/checkstyle/checkstyle/issues/13758.

How does this start the issue but doesn't completely resolve it? What is planned next?

File: src/test/java/com/puppycrawl/tools/checkstyle/checks/metrics/ClassFanOutComplexityCheckTest.java (hunk 20 above)
Comment (Lmh-java) [reply src/test/java/com/puppycrawl/tools/checkstyle/checks/metrics/ClassFanOutComplexityCheckTest.java:375]: This has nothing to do with the converter. Please see https://github.com/checkstyle/checkstyle/pull/14740#issuecomment-2047109647.

This is a part of the workaround for `ClassFanOutCheck` until #14787

TLDR;
I have to keep the following line, otherwise the CI will fail.
https://github.com/checkstyle/checkstyle/blob/7d5d4d21e3a139ec80120f02e08def8d347f9911/src/main/java/com/puppycrawl/tools/checkstyle/checks/metrics/AbstractClassCouplingCheck.java#L121
Added this test case to kill Pitest mutation for this line. After #14787 is solved, we will not need this anymore.

File: src/xdocs/property_types.xml (hunk 19 above)
Comment (Lmh-java) [reply src/xdocs/property_types.xml:676]: done

File: src/main/java/com/puppycrawl/tools/checkstyle/AbstractAutomaticBean.java (hunk 18 above)
Comment (Lmh-java) [reply src/main/java/com/puppycrawl/tools/checkstyle/AbstractAutomaticBean.java]: done

File: src/main/java/com/puppycrawl/tools/checkstyle/AbstractAutomaticBean.java (hunk 17 above)
Comment (Lmh-java) [reply src/main/java/com/puppycrawl/tools/checkstyle/AbstractAutomaticBean.java]: done

File: src/main/resources/com/puppycrawl/tools/checkstyle/meta/checks/imports/ImportOrderCheck.xml (hunk 21 above)
Comment (romani) [reply src/main/resources/com/puppycrawl/tools/checkstyle/meta/checks/imports/ImportOrderCheck.xml:43]: It will be declared breaking compatibility, I hope eclipscs will not be actually affected.

File: src/main/java/com/puppycrawl/tools/checkstyle/AbstractAutomaticBean.java [hunk 22]
```diff
@@ -313,6 +315,24 @@ public Object convert(Class type, Object value) {
...
+    /** A converter that converts a comma-separated string into an array of patterns. */
+    private static final class PatternArrayConverter implements Converter {
+
+        @SuppressWarnings("unchecked")
+        @Override
+        public Object convert(Class type, Object value) {
+            final String plainString = value.toString();
+            Pattern[] patterns = new Pattern[0];
+            if (!CommonUtil.isBlank(plainString)) {
+                final String[] patternStrings = plainString.split(COMMA_SEPARATOR);
```
Comment (rnveach) [inline src/main/java/com/puppycrawl/tools/checkstyle/AbstractAutomaticBean.java:327]: Sorry I missed this before.

https://github.com/checkstyle/checkstyle/blob/8dc065a6cbe028141ba0f863d14d95ecae4675ef/src/main/java/com/puppycrawl/tools/checkstyle/AbstractAutomaticBean.java#L393
https://github.com/checkstyle/checkstyle/blob/8dc065a6cbe028141ba0f863d14d95ecae4675ef/src/main/java/com/puppycrawl/tools/checkstyle/AbstractAutomaticBean.java#L398
This is similar to `RelaxedStringArrayConverter` and it trims the input during the split. You are not.

You can see in `testBeanConverterPatternArray` that the space remains in the regular expression. This violates your example you added in `property_types.xml` as it shows the space removed.

Originally we were using `RelaxedStringArrayConverter` for these type of instances, so we should continue the trimming it does. We need to implement trim in this method. I would expect spaces to be trimmed.

**Review by rnveach — CHANGES_REQUESTED** [review]

**Comment by rnveach** [comment]
> > Start Issue https://github.com/checkstyle/checkstyle/issues/13758.

> How does this start the issue but doesn't completely resolve it? What is planned next?

I did not get a response, but I see text was updated in first post. I am more curious now. Is this PR plan to finish the issue?

**Comment by Lmh-java** [comment]
> > > Start Issue #13758.
> 
> > How does this start the issue but doesn't completely resolve it? What is planned next?
> 
> I did not get a response, but I see text was updated in first post. I am more curious now. Is this PR plan to finish the issue?

Yes, I updated the old PR description. All these 3 steps listed now should be completed in this PR, and it should finish the issue.

File: src/main/java/com/puppycrawl/tools/checkstyle/AbstractAutomaticBean.java (hunk 22 above)
Comment (Lmh-java) [reply src/main/java/com/puppycrawl/tools/checkstyle/AbstractAutomaticBean.java]: > Sorry I missed this before.
Totally no worries. This is a great catch. I was about to ask about this, but as time went by, I forgot...

Added trim. I have also updated the new converter to have the same structure with `RelaxedStringArrayConverter`, so we can reduce one import to `Arrays`.

File: src/main/resources/com/puppycrawl/tools/checkstyle/meta/checks/imports/ImportOrderCheck.xml (hunk 21 above)
Comment (rnveach) [reply src/main/resources/com/puppycrawl/tools/checkstyle/meta/checks/imports/ImportOrderCheck.xml:43]: So we are not going to do any analysis or warning?

**Review by rnveach — APPROVED** [review]

**Comment by nrmancuso** [comment]
Github, generate site

**Comment by rnveach** [comment]
> https://checkstyle-diff-reports.s3.us-east-2.amazonaws.com/9f55855_2024115508/property_types.html#

Weird it failed to generated the fragment `Pattern[]`.

**Review by nrmancuso — APPROVED** [review]

File: src/main/resources/com/puppycrawl/tools/checkstyle/meta/checks/imports/ImportOrderCheck.xml (hunk 21 above)
Comment (romani) [reply src/main/resources/com/puppycrawl/tools/checkstyle/meta/checks/imports/ImportOrderCheck.xml:43]: yes, let move forward, and just declare breaking changes.

**Comment by Lmh-java** [comment]
@rnveach 
Could you please explain a little more on "fragment Pattern[]". I am not very sure what it means.
I can see `Pattern[]` on the site.
Pattern[] section:
image
Content section.
image

**Comment by rnveach** [comment]
https://en.wikipedia.org/wiki/URI_fragment
In the URL, it is represented by the `#` and what comes after it.

https://github.com/checkstyle/checkstyle/pull/14740#issuecomment-2088358611
The links generated by the GHA did not provide the correct fragment for the last one. My edit provided why it didn't.

**Comment by Lmh-java** [comment]
I get it right now.

So we expect the last link to be https://checkstyle-diff-reports.s3.us-east-2.amazonaws.com/9f55855_2024115508/property_types.html#Pattern.5B.5D
(with an anchor in the end)

I will open an issue for this to the main checkstyle repo.

**Comment by nrmancuso** [comment]
@Lmh-java i think we might have an issue open for it, please try to search and see if it turns up

**Comment by Lmh-java** [comment]
I didn't see an issue related to this specifically, I will try to open an issue.

**Comment by nrmancuso** [comment]
Thanks @Lmh-java !
